server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -364,7 +383,31 @@ public VolumeVO uploadVolume(UploadVolumeCmd cmd) throws ResourceAllocationExcep
         RegisterVolumePayload payload = new RegisterVolumePayload(cmd.getUrl(), cmd.getChecksum(), cmd.getFormat());
         vol.addPayload(payload);
 
-        volService.registerVolume(vol, store);
+        return registerVolume(vol, store, sizeInGB);
+    }
+
+    VolumeVO registerVolume(VolumeInfo vol, DataStore store, Long sizeInGB ) {
+        VolumeVO volume = _volsDao.findById(vol.getId());
+        try {
+            AsyncCallFuture<VolumeApiResult> future = volService.registerVolume(vol, store);
+            VolumeApiResult result = future.get();
+            if (result.isFailed()) {
+                s_logger.warn("Failed to register the volume");
+                String details = "";
+                if (result.getResult() != null && !result.getResult().isEmpty()) {
+                    details = result.getResult();
+                }
+                throw new CloudRuntimeException(details);
+            }
+            if (!Objects.equals(sizeInGB, vol.getSize())) {
+                if (sizeInGB !=null) {
+                    volume.setSize(sizeInGB * GiB_TO_BYTES);
+                    _volsDao.persist(volume);
+                }
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException(String.format("Failed to register volume due to - %s", e.getMessage()), e);
```

Comment:
We can add some info about the volume here.

Reply:
@GutoVeronezi what do you expect? Keeping in mind that the upload/registering failed, so effectively there is no volume to speak about.